[TensorExpr] Simplify TE IR before applying any transformations.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64717

This also exposed several bugs, which are fixed in this PR.

Differential Revision: [D30826408](https://our.internmc.facebook.com/intern/diff/D30826408)

Commit:
[TensorExpr] Properly construct indices for dimension with size 1 for output tensors.

Previously we incorrectly handled a case when size was 1, and thus the
value of the corresponding stride didn't play any role.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Simplify TE IR before applying any transformations."


This also exposed several bug, which are fixed in this PR.

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30826408).

Commit:
Update on "[TensorExpr] Simplify TE IR before applying any transformations."


This also exposed several bugs, which are fixed in this PR.

Differential Revision: [D30826408](https://our.internmc.facebook.com/intern/diff/D30826408)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30826408).

navahgar commented on test/cpp/tensorexpr/test_kernel.cpp:
```diff
@@ -210,7 +210,11 @@ TEST_F(Kernel, Huge) {
+  // outer loop will be parallel. If LLVM is not present, it will not be split,
+  // and to cover both of these cases we're looking for 00000000ll; in the
+  // output.
+  const std::string& verification_pattern = R"IR(# CHECK: 00000000ll;)IR";
```
nit: we don't quite have to check for that number. We just want to make sure that the compilation is successful here. So, even checking for something like `aten_relu` would suffice..

ZolotukhinM replied:
I wanted to still have a check for a number to make sure we're not truncating it or something. But if it happens to be too fragile, we can indeed just check that it compiles.

navahgar commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -731,6 +731,9 @@ bool LoopNest::computeInline(StmtPtr s) {
 bool LoopNest::computeInline(BufPtr b) {
   // If buf is used or defined in an ExternalCall, we cannot inline it
   auto buf_load_store_uses = findLoadOrStoreUses(root_stmt_);
+  if (!buf_load_store_uses.count(b)) {
```
When does this case happen? Just curious. Also, is this relevant / required for this PR?

ZolotukhinM replied:
Yes, this is required, and it happens in one of our tests where we concat empty tensors. Before I added simplifier, we had a stmt like
```
{
for (i = 0; i < 0; i++)
  a[i] = 0
}
```
With simplifier the entire block becomes empty.

Commit:
Update on "[TensorExpr] Simplify TE IR before applying any transformations."


This also exposed several bugs, which are fixed in this PR.

Differential Revision: [D30826408](https://our.internmc.facebook.com/intern/diff/D30826408)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30826408).

Commit:
rebase on "[TensorExpr] Simplify TE IR before applying any transformations."


This also exposed several bugs, which are fixed in this PR.

Differential Revision: [D30826408](https://our.internmc.facebook.com/intern/diff/D30826408)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30826408).

Commit:
Update on "[TensorExpr] Simplify TE IR before applying any transformations."


This also exposed several bugs, which are fixed in this PR.

Differential Revision: [D30826408](https://our.internmc.facebook.com/intern/diff/D30826408)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30826408).
